File: app/src/main/java/com/pluu/support/olleh/OllehWeekApi.java
package com.pluu.support.olleh;

import android.content.Context;
import android.text.TextUtils;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.json.JSONArray;
import org.json.JSONObject;

import com.pluu.support.impl.AbstractWeekApi;
import com.pluu.support.impl.ServiceConst;
import com.pluu.webtoon.R;
import com.pluu.webtoon.api.Status;
import com.pluu.webtoon.api.WebToonInfo;
import com.pluu.webtoon.api.WebToonType;

/**
 * 올레 웹툰 Week API
 * Created by PLUUSYSTEM-NEW on 2015-10-31.
 */
public class OllehWeekApi extends AbstractWeekApi {

	private static final String[] TITLE = new String[]{"월", "화", "수", "목", "금", "토", "일"};
	private static final String WEEKLY_URL = "http://webtoon.olleh.com/api/work/getWorkList.kt";
	private static final String[] WEEKLY_VALUE = {"mondayyn", "tuesdayyn", "wednesdayyn", "thursdayyn", "fridayyn", "saturdayyn", "sundayyn"};

	private JSONArray savedArray;
	private int totalSize;

	private final String adult = "adult";
	private final String novel = "novel";

	public OllehWeekApi() {
		super(TITLE);
	}

	@Override
	public ServiceConst.NAV_ITEM getNaviItem() {
		return ServiceConst.NAV_ITEM.OLLEH;
	}

	@Override
	protected int getMainTitleColor(Context context) {
		return R.color.olleh_color;
	}

	@Override
	public List<WebToonInfo> parseMain(Context context, int position) {
		ArrayList<WebToonInfo> list = new ArrayList<>();

		try {
			if (savedArray == null) {
				String response = requestApi();
				savedArray = new JSONObject(response).optJSONArray("workList");
				totalSize = savedArray.length();
			}

			WebToonInfo item;
			JSONObject obj;

			StringBuffer writerBuffer = new StringBuffer();
			String temp;
			final String optY = "Y";
			final String optN = "N";

			String checkValue = WEEKLY_VALUE[position];

			for (int i = 0; i < totalSize ; i++) {
				obj = savedArray.optJSONObject(i);
				if (!TextUtils.equals(obj.optString(checkValue), optY)) {
					continue;
				}

				item = new WebToonInfo(obj.optString("webtoonseq"));
				item.setUrl(item.getWebtoonId());
				item.setTitle(obj.optString("webtoonnm"));
				item.setImage(obj.optString("newthumbpath"));

				writerBuffer.setLength(0);
				writerBuffer.append(obj.optString("authornm1"));

				temp = obj.optString("authornm2");
				if (!TextUtils.isEmpty(temp)) {
					writerBuffer.append(", ").append(temp);
				}
				temp = obj.optString("authornm3");
				if (!TextUtils.isEmpty(temp)) {
					writerBuffer.append(", ").append(temp);
				}
				item.setWriter(writerBuffer.toString());
				item.setRate(obj.optString("totalstickercnt"));
				item.setUpdateDate(obj.optString("regdt"));
				if (optN.equals(obj.optString("endyn", optN))) {
					if (optY.equals(obj.optString("upyn", optN))) {
						// 최근 업데이트
						item.setStatus(Status.UPDATE);
					} else if (optY.equals(obj.optString("restyn", optN))){
						// 휴재
						item.setStatus(Status.BREAK);
					}
				}

				if (adult.equals(obj.optString("agefg"))) {
					item.setIsAdult(true);
				}

				if (novel.equals(obj.optString("toonfg"))) {
					item.setType(WebToonType.NOVEL);
				}

				list.add(item);
			}
		} catch (Exception e) {
			e.printStackTrace();
		}
		return list;
	}

	@Override
	public String getMethod() {
		return POST;
	}

	@Override
	public String getId() {
		return WEEKLY_URL;
	}

	@Override
	public Map<String, String> getHeaders() {
		Map<String, String> map = new HashMap<>();
		map.put("Referer", "http://webtoon.olleh.com");
		return map;
	}

	@Override
	public Map<String, String> getParams() {
		Map<String, String> map = new HashMap<>();
		map.put("mobileyn", "N");
		map.put("toonfg", "toon");
		map.put("sort", "subject");
		return map;
	}
}

File: app/src/main/java/com/pluu/support/olleh/OllehWeekApi.kt
package com.pluu.support.olleh

import android.content.Context
import com.pluu.support.impl.AbstractWeekApi
import com.pluu.support.impl.NAV_ITEM
import com.pluu.support.impl.NetworkSupportApi
import com.pluu.webtoon.item.Status
import com.pluu.webtoon.item.WebToonInfo
import org.jsoup.Jsoup
import org.jsoup.nodes.Element

/**
 * 올레 웹툰 Week API
 * Created by pluu on 2017-04-22.
 */
class OllehWeekApi(context: Context) : AbstractWeekApi(context, OllehWeekApi.TITLE) {
    override val naviItem: NAV_ITEM = NAV_ITEM.OLLEH

    override fun parseMain(position: Int): List<WebToonInfo> {
        val weekly = try {
            Jsoup.parse(requestApi()).select("div[class=list week_all] .inner")
        } catch (e: Exception) {
            e.printStackTrace()
            return emptyList()
        }

        val dataPosition = weekly.select("h4")
                .indexOfFirst { it.text() == TITLE[position] }

        if (dataPosition == -1) return emptyList()

        return weekly[dataPosition].select("li")
                .mapNotNull { transform(it) }
                .toList()
    }

    private fun transform(it: Element): WebToonInfo? {
        val directLink = it.select(".link").attr("href")
        val toonId = "(?<=worksseq=).+".toRegex().find(directLink) ?: return null
        return WebToonInfo(toonId.value).apply {
            val info = it.select(".info")
            title = info.select("strong").text()
            image = it.select(".thumb img").attr("src")

            if (info.select("ico_up").isNotEmpty()) {
                // 최근 업데이트트
                status = Status.UPDATE
            } else if (info.select("ico_break").isNotEmpty()) {
                // 휴재
                status = Status.BREAK
            }

            isAdult = info.select("ico_adult").isNotEmpty()
            link = directLink
        }
    }

    override val url = "https://www.myktoon.com/web/webtoon/works_list.kt"

    override val method: String = NetworkSupportApi.GET

    override val headers: Map<String, String> = emptyMap()

    override val params: Map<String, String> = emptyMap()

    companion object {
        private val TITLE = arrayOf("월", "화", "수", "목", "금", "토", "일")
    }
}
